xkafka/producer.go

```diff
@@ -0,0 +1,194 @@
+package xkafka
+
+import (
+	"context"
+	"strings"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/pkg/errors"
+
+	"github.com/gojekfarm/xtools/xkafka/internal"
+)
+
+// ProducerFunc is a function that returns a KafkaProducer.
+type ProducerFunc func(cfg *kafka.ConfigMap) (internal.ProducerClient, error)
+
+func (pf ProducerFunc) apply(o *options) { o.producerFn = pf }
+
+// DefaultProducerFunc is the default producer function that initialises
+// a new confluent-kafka-go/kafka.Producer.
+func DefaultProducerFunc(cfg *kafka.ConfigMap) (internal.ProducerClient, error) {
+	return kafka.NewProducer(cfg)
+}
+
+// Producer manages the production of messages to kafka topics.
+// It provides both synchronous and asynchronous publish methods
+// and a channel to stream delivery events.
+type Producer struct {
+	config              options
+	kafka               internal.ProducerClient
+	events              chan kafka.Event
+	delivery            chan *Message
+	middlewares         []middleware
+	wrappedPublish      Handler
+	wrappedAsyncPublish Handler
+}
+
+// NewProducer creates a new Producer.
+func NewProducer(opts ...Option) (*Producer, error) {
+	cfg := defaultProducerOptions()
+
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+
+	_ = cfg.configMap.SetKey("bootstrap.servers", strings.Join(cfg.brokers, ","))
+
+	producer, err := cfg.producerFn(&cfg.configMap)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Producer{
+		config:   cfg,
+		kafka:    producer,
+		delivery: make(chan *Message),
+		events:   producer.Events(),
+	}
+
+	p.wrappedPublish = p.publish()
+	p.wrappedAsyncPublish = p.asyncPublish()
+
+	go p.start()
```

Comment:
`Start` and `Stop` should be exposed publicly IMO. Gives a user more fine grain control.